### PR TITLE
[RBAC] BREAKING CHANGE: fix issue #11883 az role assignment create scope default to subscription if empty

### PR DESCRIPTION
**Pull request by arrownj**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/azure-cli/HISTORY.rst [hunk 1]
```diff
@@ -3,6 +3,10 @@
 Release History
 ===============
 
+**RBAC**
+
+* [BREAKING CHANGE] Fix issue #11883 az role assignment create scope default to subscription if empty
```
Comment (jiasli) [inline src/azure-cli/HISTORY.rst]: Please follow the history convention like:

> Fix #11883: `az role assignment create`: scope defaults to subscription if not specified

**Comment by jiasli** [comment]
Why the behavior of `az role assignment list` is also changed? There doesn't seem to be any security concern in this command.

Also, it is also a breaking change to force one of `--all`, `--scope` and `-resource-group`, which needs to go to the history.

File: src/azure-cli/azure/cli/command_modules/role/_validators.py [hunk 2]
```diff
@@ -91,6 +91,18 @@ def process_assignment_namespace(cmd, namespace):  # pylint: disable=unused-argu
...
+        raise CLIError('usage error: please specify at least one of "--scope" and "--resource-group".')
+
+
+def process_list_assignment_namespace(cmd, namespace):  # pylint: disable=unused-argument
+    resource_group = namespace.resource_group_name
+    if namespace.scope and resource_group and getattr(resource_group, 'is_default', None):
+        namespace.resource_group_name = None  # drop configured defaults
+
+    if not namespace.show_all and not namespace.scope and not namespace.resource_group_name:
+        raise CLIError('usage error: please specify at least one of "--all", "--scope" and "--resource-group".')
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/_validators.py]: We usually use single quotes in error messages.

Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/_validators.py]: Why the behavior of `az role assignment list` is also changed? There doesn't seem to be any security concern in this command.

Also, it is also a breaking change to force one of `--all`, `--scope` and `-resource-group`, which needs to go to the history.

File: src/azure-cli/azure/cli/command_modules/role/_validators.py [hunk 3]
```diff
@@ -91,6 +91,18 @@ def process_assignment_namespace(cmd, namespace):  # pylint: disable=unused-argu
...
         namespace.resource_group_name = None  # drop configured defaults
 
+    if not namespace.scope and not namespace.resource_group_name:
+        raise CLIError('usage error: please specify at least one of "--scope" and "--resource-group".')
+
+
+def process_list_assignment_namespace(cmd, namespace):  # pylint: disable=unused-argument
+    resource_group = namespace.resource_group_name
+    if namespace.scope and resource_group and getattr(resource_group, 'is_default', None):
+        namespace.resource_group_name = None  # drop configured defaults
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/role/_validators.py]: This is a duplicate of L90-L92. I'd suggested to extract it to a function like `drop_default_group`.

File: src/azure-cli/azure/cli/command_modules/role/_validators.py (hunk 2 above)
Comment (arrownj) [reply src/azure-cli/azure/cli/command_modules/role/_validators.py]: I searched similar error message in our code base, seems we usually use nothing in such case. The error message will be `at least one of --scope and --resource-group` in the new PR.

**Comment by jiasli** [comment]
I think the security concern in the original issue #11883 is when `--scope` is **specified but with an empty argument**: `--scope ""`. Therefore, we don't need to change the behavior (fallback to subscription scope) when `--scope` is **not specified**.

We can error out here: https://github.com/Azure/azure-cli/blob/85f04e71dc4babd1bad7cbd4e0da37080ca4fb29/src/azure-cli/azure/cli/command_modules/role/custom.py#L493

The error can be something like:

```py
    if scope == "":
        raise CLIError('Invalid scope. Please use --help to view the valid format.')
```

One interesting thing I noticed is that the [role assignment REST API](https://docs.microsoft.com/en-us/azure/role-based-access-control/role-assignments-rest) does support empty scope. In that case, the create call is

```http
PUT https://management.azure.com/providers/Microsoft.Authorization/roleAssignments/a5b85ff6-2fbe-454d-bbc3-6a90f9c54d62?api-version=2018-09-01-preview
```

And the response is 
```json
  {
    "canDelegate": null,
    "id": "/providers/Microsoft.Authorization/roleAssignments/a5b85ff6-2fbe-454d-bbc3-6a90f9c54d62",
    "name": "a5b85ff6-2fbe-454d-bbc3-6a90f9c54d62",
    "principalId": "",
    "principalName": "",
    "principalType": "User",
    "roleDefinitionId": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c",
    "roleDefinitionName": "Contributor",
    "scope": "/",
    "type": "Microsoft.Authorization/roleAssignments"
  }
```
Notice `"scope": "/"`. But this behavior is not documented. 

Anyway, let's error out for empty scope `""` for now and discuss it furtherly later.

**Pull request closed without merging**